utils/validation.go

```diff
@@ -0,0 +1,130 @@
+// Utils for Http Rest Request validation and decoding
+
+/*
+Copyright © 2019, 2020 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"net/url"
+
+	"github.com/asaskevich/govalidator"
+	"github.com/gorilla/schema"
+)
+
+var decoder = schema.NewDecoder()
+
+// DecodeValidRequest validates input maps (From Query.URL, or decoded Json Body) against template and returns typed structure
+// srcs can be list of either map[string]interface{} or map[string][]string
+func DecodeValidRequest(dst interface{}, temp map[string]interface{}, srcs ...interface{}) error {
+	input := mergeMapsT(srcs...)
+	sm := stringsMap(input)
+	// add invalid key to force non keyed validator to run
+	input[""] = ""
+	// validate generic map to report correct type errors
+	_, err := govalidator.ValidateMap(input, temp)
+	if err != nil {
+		return err
+	}
+	// convert valid map to type safe struct
+	err = decoder.Decode(dst, sm)
+	return err
+}
+
+// mergeMapsT merges multiple maps of various types, trying to convert them
+// srcs can be;
+// map[string]interface{} (for example body dedoded from json)
+// map[string][]string  (for example r.URL.Query() or r.Form (as url.Values))
+func mergeMapsT(srcs ...interface{}) map[string]interface{} {
```

Comment:
we might probably move all maps-related functions into new file, WDYT?

Reply:
Good idea, it looks better this way.